python_modules/dagster/dagster/daemon/sensor.py

```diff
@@ -178,32 +229,56 @@ def execute_sensor_iteration_loop(instance, workspace, logger, until=None):
     """
     workspace_loaded_time = pendulum.now("UTC").timestamp()
 
-    workspace_iteration = 0
-    start_time = pendulum.now("UTC").timestamp()
-    while True:
-        start_time = pendulum.now("UTC").timestamp()
-        if until and start_time >= until:
-            # provide a way of organically ending the loop to support test environment
-            break
+    max_workers = int(os.getenv("DAGSTER_SENSOR_DAEMON_THREADPOOL_WORKERS", "3"))
 
-        if start_time - workspace_loaded_time > RELOAD_WORKSPACE:
-            workspace.cleanup()
-            workspace_loaded_time = pendulum.now("UTC").timestamp()
-            workspace_iteration = 0
+    with concurrent.futures.ThreadPoolExecutor(
+        max_workers=max_workers,
+        thread_name_prefix="sensor_daemon_worker",
+    ) as executor:
```

Comment:
We should only instantiate this once, probably during daemon initialization? Would need to thread it down to here.

Reply:
execute_sensor_iteration_loop is called exactly once in the sensor daemon (SensorDaemon is not an IntervalDaemon). So this is already getting instantiated exactly once in the sensor daemon

unless you mean sharing the executor across multiple daemons (scheduler/sensor/run queue/backfill/etc.), which is an interesting idea with some tradeoffs.

The pros are that we could have one executor with the ideal number of threads for the process, and when one daemon isn't doing much the other daemons would be able to get full utilization out of the threadpool

The cons are that there would be a risk of all other daemons falling behind when one daemon kicks off a bunch of threads (i'm imagining the run queue daemon will get a threadpool soon too - so imagining a bunch of runs getting launched and the sensor daemons falling behind and timing out)